src/Core/Content/MailTemplate/Service/MailService.php

```diff
@@ -140,32 +140,31 @@ public function send(array $data, Context $context, array $templateData = []): ?
         $senderEmail = $this->getSender($data, $salesChannelId);
 
         $contents = $this->buildContents($data, $salesChannel);
-        foreach ($contents as $index => $template) {
-            try {
-                if (isset($data['testMode']) && (bool) $data['testMode'] === true) {
-                    $this->templateRenderer->enableTestMode();
-                }
 
+        try {
+            if (isset($data['testMode']) && (bool) $data['testMode'] === true) {
+                $this->templateRenderer->enableTestMode();
+            }
+            $data['subject'] = $this->templateRenderer->render($data['subject'], $templateData, $context);
+            $data['senderName'] = $this->templateRenderer->render($data['senderName'], $templateData, $context);
+            foreach ($contents as $index => $template) {
                 $contents[$index] = $this->templateRenderer->render($template, $templateData, $context);
-                $data['subject'] = $this->templateRenderer->render($data['subject'], $templateData, $context);
-                $data['senderName'] = $this->templateRenderer->render($data['senderName'], $templateData, $context);
-
-                if (isset($data['testMode']) && (bool) $data['testMode'] === true) {
-                    $this->templateRenderer->disableTestMode();
-                }
-            } catch (\Exception $e) {
-                $this->logger->error(
-                    "Could not render Mail-Template with error message:\n"
-                    . $e->getMessage() . "\n"
-                    . 'Error Code:' . $e->getCode() . "\n"
-                    . 'Template source:'
-                    . $template . "\n"
-                    . "Template data: \n"
-                    . json_encode($templateData) . "\n"
-                );
-
-                return null;
             }
+            if (isset($data['testMode']) && (bool) $data['testMode'] === true) {
+                $this->templateRenderer->disableTestMode();
+            }
+        } catch (\Exception $e) {
+            $this->logger->error(
+                "Could not render Mail-Template with error message:\n"
+                . $e->getMessage() . "\n"
+                . 'Error Code:' . $e->getCode() . "\n"
+                . 'Template source:'
+                . $template . "\n"
```

Comment:
This will not be defined with the new position of the `foreach`: https://github.com/shopware/platform/pull/1436/checks?check_run_id=1280180341#step:7:300

Please implement the change in a way that doesn't affect the exception message.

Reply:
Hi @philipgatzka, thanks for the review. I fixed the undefined variable. Errors in the template rendering for subject/sender should now be logged correctly, in the old code they were logged as errors in body rendering.

Reply:
There seems to be a problem with the unit tests in a part I did not touch.

Reply:
Yes, the failing test is indeed unrelated. Thanks for fixing the undefined variable problem, the other part still remains though - please make sure the exception message isn't altered. As far as I can tell, you'll now see a message containing only the recipients name for example, instead of the whole template (which could be useful for conext). You could for example implement a test which checks whether the exception message is complete and then implement your changes.

Reply:
I'm not sure I understand. Do you mean I should log the mail body even if it did not cause the error, because the body does contain more content, which could help identifying the mail as a whole?

## The current Version:

If the renderer throws here, the template, which caused the error will appear in the message.
https://github.com/shopware/platform/blob/d52d6b32730c9bee2fcdc8fa8a75b409bee621fc/src/Core/Content/MailTemplate/Service/MailService.php#L149

If the renderer throws here, the template, which is actually just the subject, which caused the error will not appear in the message. The mail body, which rendered fine, will appear in the message.
https://github.com/shopware/platform/blob/d52d6b32730c9bee2fcdc8fa8a75b409bee621fc/src/Core/Content/MailTemplate/Service/MailService.php#L150

## My Version:

`$template = $data['senderName'];` will cause the actual template currently being rendered to be assigned to the variable of the error message. So the error message, should contain the string, which actually caused the templateRenderer to throw.

Reply:
Ahh yes, thanks for the explanation. Didn't get it at first :smile: Alright :+1: